File: api/serenity/watcher/rules.py
"""Password rules shared with the browser (web/src/features/breaches/rules.ts).

Both implementations are checked against shared/test-vectors/watch.json.
"""

import math
import re
from datetime import datetime, timedelta

MIN_LENGTH = 12
MIN_BITS = 60
MIN_DISTINCT = 5
OLD_AFTER = timedelta(days=365)

_POOLS = (
    (re.compile(r"[a-z]"), 26),
    (re.compile(r"[A-Z]"), 26),
    (re.compile(r"[0-9]"), 10),
    (re.compile(r"[^a-zA-Z0-9]"), 33),
)


def strength_bits(password: str) -> float:
    """Brute-force size for the character classes used (an upper bound, not a guess estimate)."""
    pool = sum(size for pattern, size in _POOLS if pattern.search(password))
    return len(password) * math.log2(pool) if pool else 0.0


def is_weak(password: str) -> bool:
    """Too short, too small a character space, or too repetitive. Empty passwords are skipped."""
    if not password:
        return False
    return (
        len(password) < MIN_LENGTH
        or strength_bits(password) < MIN_BITS
        or len(set(password)) < MIN_DISTINCT
    )


def is_old(changed_at: datetime | None, now: datetime) -> bool:
    return changed_at is not None and now - changed_at > OLD_AFTER


def parse_date(value: object) -> datetime | None:
    if not isinstance(value, str) or not value:
        return None
    try:
        return datetime.fromisoformat(value.replace("Z", "+00:00"))
    except ValueError:
        return None

File: web/src/features/breaches/rules.ts
/**
 * Password rules shared with the server (api/serenity/watcher/rules.py), checked against
 * shared/test-vectors/watch.json.
 */

export const MIN_LENGTH = 12;
export const MIN_BITS = 60;
export const MIN_DISTINCT = 5;
export const OLD_AFTER_DAYS = 365;

const POOLS: [RegExp, number][] = [
  [/[a-z]/, 26],
  [/[A-Z]/, 26],
  [/[0-9]/, 10],
  [/[^a-zA-Z0-9]/, 33],
];

/** Code points, like Python len(). */
function length(text: string): number {
  return Array.from(text).length;
}

export function strengthBits(password: string): number {
  const pool = POOLS.reduce(
    (sum, [pattern, size]) => (pattern.test(password) ? sum + size : sum),
    0,
  );
  return pool ? length(password) * Math.log2(pool) : 0;
}

export function isWeak(password: string): boolean {
  if (!password) return false;
  return (
    length(password) < MIN_LENGTH ||
    strengthBits(password) < MIN_BITS ||
    new Set(Array.from(password)).size < MIN_DISTINCT
  );
}

export function isOld(changedAt: Date | null, now: Date): boolean {
  return changedAt !== null && now.getTime() - changedAt.getTime() > OLD_AFTER_DAYS * 86_400_000;
}

export function parseDate(value: unknown): Date | null {
  if (typeof value !== "string" || !value) return null;
  const date = new Date(value);
  return Number.isNaN(date.getTime()) ? null : date;
}
